File: kp/service/vm.py
from proxmoxer import ProxmoxAPI

from kp import config
from kp.client.pve import PveApi
from kp.error import *


class VmService:
    @staticmethod
    def systemctl_daemon_reload(api: ProxmoxAPI,
                                node: str,
                                vm_id: str):
        cmd = "systemctl daemon-reload".split()
        return PveApi.exec(api, node, vm_id, cmd)

    @staticmethod
    def update_containerd_config(api: ProxmoxAPI,
                                 node: str,
                                 vm_id: str,
                                 content=config.CONTAINERD_CONFIG):
        cmd = "mkdir -p /etc/containerd".split()
        PveApi.exec(api, node, vm_id, cmd)
        PveApi.write_file(api, node, vm_id, "/etc/containerd/config.toml", content)

    @staticmethod
    def restart_containerd(api: ProxmoxAPI,
                           node: str,
                           vm_id: str):
        cmd = ["systemctl", "restart", "containerd"]
        return PveApi.exec(api, node, vm_id, cmd)

    @staticmethod
    def restart_kubelet(api: ProxmoxAPI,
                        node: str,
                        vm_id: str):
        cmd = ["systemctl", "restart", "kubelet"]
        return PveApi.exec(api, node, vm_id, cmd)

    @staticmethod
    def kubeadm_reset(api: ProxmoxAPI,
                      node: str,
                      vm_id: str, cmd=["kubeadm", "reset", "-f"]):
        return PveApi.exec(api, node, vm_id, cmd)

File: model/vm.go
package model

import (
	"context"
	"fmt"
	"strings"

	"github.com/luthermonson/go-proxmox"
	"github.com/tuana9a/kp/constants"
	"github.com/tuana9a/kp/templates"
)

type VirtualMachineV2 struct {
	V1     *proxmox.VirtualMachine
	Client *proxmox.Client
}

func (vm *VirtualMachineV2) AgentFileWrite(ctx context.Context, location string, content string) error {
	err := vm.Client.Post(ctx, fmt.Sprintf("/nodes/%s/qemu/%d/agent/file-write", vm.V1.Node, vm.V1.VMID), map[string]string{"file": location, "content": content}, nil)
	return err
}

func (vm *VirtualMachineV2) AgentFileRead(ctx context.Context, location string, content *string) error {
	err := vm.Client.Get(ctx, fmt.Sprintf("/nodes/%s/qemu/%d/agent/file-read?file=%s", vm.V1.Node, vm.V1.VMID, location), &content)
	return err
}

func (vm *VirtualMachineV2) EnsureContainerdConfig(ctx context.Context) error {
	pid, _ := vm.V1.AgentExec(ctx, []string{"mkdir", "-p", "/etc/containerd"}, "")
	status, err := vm.V1.WaitForAgentExecExit(ctx, pid, 5)
	if err != nil {
		return err
	}
	if status.ExitCode != 0 {
		// TODO
	}
	vm.AgentFileWrite(ctx, constants.ContainerdConfigPath, templates.ContainerdConfig)
	return nil
}

func (vm *VirtualMachineV2) RestartContainerd(ctx context.Context) error {
	pid, _ := vm.V1.AgentExec(ctx, []string{"systemctl", "restart", "containerd"}, "")
	status, err := vm.V1.WaitForAgentExecExit(ctx, pid, 10)
	if err != nil {
		return err
	}
	if status.ExitCode != 0 {
		// TODO
	}
	return nil
}

func (vm *VirtualMachineV2) CreateWorkerJoinCommand(ctx context.Context) ([]string, error) {
	pid, _ := vm.V1.AgentExec(ctx, []string{"kubeadm", "token", "create", "--print-join-command"}, "")
	status, err := vm.V1.WaitForAgentExecExit(ctx, pid, 5)
	if err != nil {
		return nil, err
	}
	if status.ExitCode != 0 {
		// TODO
	}
	cmd := strings.Split(strings.Trim(status.OutData, " \n"), " ")
	return cmd, nil
}

func (vm *VirtualMachineV2) CreateControlPlaneJoinCommand(ctx context.Context) ([]string, error) {
	pid, _ := vm.V1.AgentExec(ctx, []string{"kubeadm", "token", "create", "--print-join-command"}, "")
	status, err := vm.V1.WaitForAgentExecExit(ctx, pid, 5)
	if err != nil {
		return nil, err
	}
	if status.ExitCode != 0 {
		// TODO
	}
	cmd := append(strings.Split(strings.Trim(status.OutData, " \n"), " "), "--control-plane")
	return cmd, nil
}

func (vm *VirtualMachineV2) WaitForCloudInit(ctx context.Context) error {
	pid, _ := vm.V1.AgentExec(ctx, constants.CloudInitWaitCmd, "")
	fmt.Println("Wait for cloud-init", "pid", pid)
	status, err := vm.V1.WaitForAgentExecExit(ctx, pid, 15*60)
	if err != nil {
		// 500 Agent error: Invalid parameter 'pid'
		fmt.Println("Wait for cloud-init has error", err)
		// But we can continue
		return nil
	}
	if status.ExitCode != 0 {
		fmt.Println("Wait for cloud init did not exit successfully")
		fmt.Println("stdout")
		fmt.Println(status.OutData)
		fmt.Println("stderr")
		fmt.Println(status.ErrData)
	}
	return nil
}
